README.md
[![Build Status][bi]][bl]
[![MIT License][li]][ll]
[![Twitter][ti]][tl]

<p align="center"><a href="https://www.osano.com/features/consent-management?utm_campaign=Open%20Source%20Cookie%20Consent&utm_source=github"><img width="460" src="images/cookie-consent.png"></a></p>

## What is Cookie Consent?

[Cookie Consent][cl] is a lightweight JavaScript plugin for alerting users about the use of cookies on your website.

It is designed to help you quickly comply with the EU Cookie Law, CCPA, GDPR and other privacy laws. We made it fast, free, and relatively painless.

Cookie Consent is seen over 2 BILLION times every month and is used on millions of sites, making this by far the most popular consent project on the internet.

We welcome community contributions and actively review pull requests.


## Hosted Zero Config Consent

Our parent company Osano actively maintains this open source project. Osano is a B-corp and is backed by millions of dollars in venture capital to create data privacy tools for the world. Our mission is to create transparency in data, please consider [joining our team](https://www.osano.com/company/careers).

Ensuring that your website is compliant with various laws can be complicated. To effectively use **any** open source consent manager, you will need to do GeoIP lookups, adjust the consent types based on visitor location, callback and save consents in a database, and create callbacks to load scripts after consent is granted.

Osano offers a completely [free hosted consent management platform][cl], with additional features. If you have a high traffic website or need additional privacy tools, Osano offers paid plans as well. **Unless you specifically need the open source tool, most website owners will be better served by the hosted version.** If you use Osano for your business, please support continued development by purchasing a [paid plan](https://www.osano.com/plans?utm_campaign=Open%20Source%20Cookie%20Consent&utm_source=github).

The hosted version of the cookie consent manager provides additional capabilities such as:

* Multi-lingual, translated into 38 languages and growing. Visitors automatically see the dialog in their browser or OS preferred language.
* Consents are stored automatically on the blockchain. REST API calls are available to developers to search and browse the consent records.
* 3rd party script blocking and loading does not require callbacks but is instead configurable from a dashboard providing developers with control over what the marketing team can enable or disable.
* Hosted on AWS Cloudfront with edges around the globe for 

## Basic Use

With version 4.0 you only need to attach the script as we've bundled everything together now. The initialization style has changed as have the callbacks (they're gone). Please see the text below to get started.  Then, take a look at updated API via the [docs][dl].

#### Module
```
import CC from "CookieConsent"
// or
const CC = require( "CookieConsent" )
```

#### Classic
```
const CC = window.CookieConsent
```

##### Initialization:
```
const cc = new CC({
  //...options,
  type   : "categories"
})
```

##### Lifecycle hooks, are now events: 
```
cc.on( "initialized", ( ...args ) => console.log( args ) )
cc.on( "error", console.error )
cc.on( "popupOpened", () => console.log( "Popup Open" ) )
cc.on( "popupClosed", () => console.log( "Popup Closed" ) )
cc.on( "revokeChoice", () => console.log( "Popup Reset" ) )
cc.on( "statusChanged", ( ...args ) => console.log( args ) )
```


## Version 4.0
Lots of updates & some breaking changes... but they're all for the better, we promise!

## Version 3.1

Reflects the ownership change of the Cookie Consent project. Now actively maintained by:

- @arlogilbert
- @L0key
- @pgoforth

## Version 3.0

Version 3.0 is a complete rewrite from version 2. The most substantial new features are:

- Ability to GeoLocate and only show the add-on to people in the relevant countries
- Callback hooks for showing/accepting/revoking the banner
- Support for different types of compliance, giving you the flexibility to obey even the strictest cookie laws
- Easy no-fuss themes and customisable styles

## Installation

The easiest way to get up and running is to use our [wizard][dll].

You can also install this project through [npm](https://www.npmjs.com/package/cookieconsent):

```sh
npm install cookieconsent
```

Or through [Yarn](https://yarnpkg.com/en/package/cookieconsent):

```sh
yarn add cookieconsent@3
```

Or through [Bower](https://bower.io/):

```sh
bower install cookieconsent
```

Or via a jsDelivr:

```html
<script src="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.js"></script>
```

## Documentation

See our [full documentation][dl].

## Contributing

Feel free to improve the plugin and send us a pull request.

The easiest way to develop is to host the files with a local webserver. e.g.

```sh
python -m SimpleHTTPServer
```

We use Babel, Terser, and PostCSS to compile the SCSS and minify the JavaScript. You can run a build with:

```sh
npm run build
```

or

```sh
yarn run build
```

## Legal Mumbo Jumbo

Osano is a registered trademark of Osano, Inc.

Nothing on the Osano website, platform, services, or software, nor any portion thereof constitutes actual legal or regulatory advice, opinion, or recommendation by Osano, Inc. or Osano International Compliance Services LTD. 

**If legal assistance is required, users should seek the services of an attorney.**

## Export Control

This distribution includes cryptographic software. The country in which you
currently reside may have restrictions on the import, possession, use, and/or
re-export to another country, of encryption software. BEFORE using any
encryption software, please check your country's laws, regulations and
policies concerning the import, possession, or use, and re-export of encryption
software, to see if this is permitted. See <http://www.wassenaar.org/> for more
information.

The U.S. Government Department of Commerce, Bureau of Industry and Security
(BIS), has classified this software as Export Commodity Control Number (ECCN)
5D002.C.1, which includes information security software using or performing
cryptographic functions with asymmetric algorithms. The form and manner of this
Apache Software Foundation distribution makes it eligible for export under the
License Exception ENC Technology Software Unrestricted (TSU) exception (see the
BIS Export Administration Regulations, Section 740.13) for both object code and
source code.

[li]: https://img.shields.io/badge/license-MIT-brightgreen.svg
[ll]: LICENSE
[bl]: https://travis-ci.org/osano/cookieconsent
[bi]: https://travis-ci.org/osano/cookieconsent.svg?branch=master
[dl]: https://www.osano.com/cookieconsent/documentation/
[dll]: https://www.osano.com/features/consent-management?utm_campaign=Open%20Source%20Cookie%20Consent&utm_source=github
[cl]: https://www.osano.com/features/consent-management?utm_campaign=Open%20Source%20Cookie%20Consent&utm_source=github
[ti]: https://img.shields.io/twitter/url/https/osanoatx.svg?style=social
[tl]: https://twitter.com/osano


## Custom Work (Mashbo)

In webpack.prod.js, amend the minify config to allow console log, debugger and dont mangle for prod. then run `yarn build && cp build/cookieconsent.min.js ../../hub-of-hope/hub-of-hope/public/js/cookieconsent.min.js`, change directories to where it is going. For actual production, change webpack.prod.js back.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [4.0.0] - Release Date

### Added
- Watches key press for allow or deny, default keys are 13 (Enter) and 27 (Esc) respectively [#571, #487]
- Categories: 'Uncategorized', 'Essential', 'Analytics', 'Marketing', 'Personalization' `{ /*...options*/, type: "categories" }`
- Events: 'initialized', 'error', 'popupOpened', 'popupClosed', 'revokeChoice', 'statusChanged' `cc.on( 'error', console.error.bind( console ) )`
- Moved to Webpack 4.0: This move integrates the CSS(SCSS) into the JS so there is only one file needed from the CDN [#558]
- Added the link click event watcher that will dismiss the pop-up if option `dismissOnLinkClick: (false)` is set to true [#503]
- Added `cookie` option `sameSite`.
  Supports value 'Lax', 'Strict' and 'None', default is 'Lax', disable using `cookie.sameSite: ''`
- Added geolocation service `freegeoip` and set as default. (15,000 queries per hour)
– Added more countries having cookie laws:
  - Australia
  - Brasil
  - Chile
  - India
  - Japan
  - South Korea
  - New Zealand
  - Russia
  - Thailand
  - United States

### Security
- Updated all dependencies to latest stable releases
- `npm audit`

### Fixed
- Throws appropriate message if there is no container for the popup to attach to [#579]
- Opt-out changes cookie value appropriately [#560] ( we're not destroying them as they fall under the 'essential' category)
- 'initialise' callback now 'intialized' event that fires before any user actions are possible [#474]
- Removed `event.path` from the code base [#502]

### Refactor:
- Started move to ES6
- Distribute utilities
- Set constants
- Event emitter added
- Removed utilities that have built-in JS functionality solutions
- Content options mostly refactored into nested properties with flattened
  text replacers `{{buttons.allow}}` and so on.  
  **Default Content options:**
  ```
  content: {
    header: 'Cookies used on the website!',
    message: 'This website uses cookies to ensure you get the best experience on our website.',
    policy: 'Cookie Policy',
    link: {
      text: 'Learn more',
      title: 'Learn more about cookies',
      href: 'https://www.cookiesandyou.com',
      target: '_blank'
    },
    buttons: {
      allow: 'Allow cookies',
      close: '&#x274c',
      deny: 'Decline',
      dismiss: 'Got it!',
      save: 'Save',
      tooltip: '^'
    },
    categories: {
      uncategorized: {
        label: 'Uncategorized',
        description: 'This is the category for cookies that don\'t fit any of the other categories'
      },
      essential: {
        label: 'Essential',
        description: 'This is the category for cookies that are essential'
      },
      personalization: {
        label: 'Personalization',
        description: 'This is the category for cookies that are used for personalization'
      },
      analytics: {
        label: 'Analytics',
        description: 'This is the category for tracking cookies'
      },
      marketing: {
        label: 'Marketing',
        description: 'This is the category for marketing cookies'
      },
    },
  }
  ```

### Deprecated:
- Callbacks: `onInitialise`, `onPopupOpen`, `onPopupClose`, `onStatusChange`, `onRevokeChoice`

### Where can you help?
- Move storage to localStorage [#455]
- Multiligual support [#570]
- Allow for custom categories
- More lifecycle events

## [3.1.1] - 2019-05-22

### Security

- Updated all dependencies to latest stable releases

### Fixed

- Expiration date future calculation error
- Closing with touch out did not work when cookie consent was displayed on iPhone Safari and dismissOnWindowClick is enabled

### Random

- Insite branding and domains replaced with Osano branding and domains

## [3.1.0] - 2018-07-17

### Added

- Secure cookies (must use the `secure` flag) in the `cookie` option
- Added a onNoCookieLaw hook
- Added a few `content` options like `policy` and `target`
- Added `dismissOnWindowClick` option (with an `ignoreClickFrom`) option which blacklists any clicks

### Fixed

- Icons (plus anything else) now allowed in buttons
- Dont call location service again if user has already chosen their privacy level
- Revoke option bug (fixed bug where revoke option wouldn't show on refresh after user has picked choice)

### Random

- Lowered console log level from error to warning when location service fails
- Removed non-standard X-Requested-By header from XxmHttpRequest
- Removed Maxmind as a default service, but you can still add it yourself (you need to whitelist a domain with maxmind)

### Deprecated

- Removed deprecated FreeGeoIP

## [3.0.2] - 2016-10-17

### Added

- Accessibility
  - :focus effects
  - ARIA rules for accessibility
- GoogleOff/GoogleOn tags for all content inside the banner. This prevents Google registering CookieConsent text as site content.

## [3.0.1] - 2016-09-21

### Added

- Changelog
- Application written to support:
  - geolocation
  - callback hooks
  - different consent types
  - new and easy to customize stylings


src/utils/index.spec.js
"use strict"

import {
  getCookie,
  setCookie,
  interpolateString,
  flattenObject,
  throttle,
  hash,
  normalizeHex,
  getContrast,
  getLuminance,
  getHoverColor,
  traverseDOMPath,
  addCustomStylesheet,
  isValidStatus,
  isPlainObject,
  isMobile,
  getScript,
  makeAsyncRequest
} from "./"

describe( "Utils::Index", () => {
  describe( "All exports are available" , () => {
    test( "getCookie", () => expect( getCookie ).toBeInstanceOf( Function ) )
    test( "setCookie", () => expect( setCookie ).toBeInstanceOf( Function ) )
    test( "interpolateString", () => expect( interpolateString ).toBeInstanceOf( Function ) )
    test( "flattenObject", () => expect( flattenObject ).toBeInstanceOf( Function ) )
    test( "throttle", () => expect( throttle ).toBeInstanceOf( Function ) )
    test( "hash", () => expect( hash ).toBeInstanceOf( Function ) )
    test( "normalizeHex", () => expect( normalizeHex ).toBeInstanceOf( Function ) )
    test( "getContrast", () => expect( getContrast ).toBeInstanceOf( Function ) )
    test( "getLuminance", () => expect( getLuminance ).toBeInstanceOf( Function ) )
    test( "getHoverColor", () => expect( getHoverColor ).toBeInstanceOf( Function ) )
    test( "traverseDOMPath", () => expect( traverseDOMPath ).toBeInstanceOf( Function ) )
    test( "addCustomStylesheet", () => expect( addCustomStylesheet ).toBeInstanceOf( Function ) )
    test( "isValidStatus", () => expect( isValidStatus ).toBeInstanceOf( Function ) )
    test( "isPlainObject", () => expect( isPlainObject ).toBeInstanceOf( Function ) )
    test( "isMobile", () => expect( isMobile ).toBeInstanceOf( Function ) )
    test( "getScript", () => expect( getScript ).toBeInstanceOf( Function ) )
    test( "makeAsyncRequest", () => expect( makeAsyncRequest ).toBeInstanceOf( Function ) )
  })
})

webpack.common.js
"use strict"

const join = require( "path" ).join
const common = require( "./webpack.common.js" )
const exp = Object.assign({}, common, {
    entry: join( __dirname, 'src', 'index.js' ),
    module: {
      rules: [
        {
          test   : /\.js?$/,
          loader : "babel-loader",
          options: {
            root    : __dirname,
            rootMode: "upward-optional"
          }
        },
        {
          test: /\.scss?$/,
          use : [
            "style-loader", // creates style nodes from JS strings
            "css-loader", // translates CSS into CommonJS
            {
              loader: "postcss-loader", // process CSS with PostCSS
              options: {
                postcssOptions: {
                  plugins: [
                    "postcss-flexbugs-fixes",
                    [
                      "postcss-preset-env",
                      {
                        "autoprefixer": {
                          "flexbox": "no-2009",
                          "grid": "autoplace"
                        },
                        "stage": 3,
                        "features": {
                          "custom-properties": false
                        }
                      },
                    ],
                  ],
                },
              }
            },
            "sass-loader" // compiles Sass to CSS, using Node Sass by default
          ]
        }
      ]
    }
  })
  module.exports = exp

src/utils/index.js
"use strict"

import * as cookie from "./cookie"
export const getCookie = cookie.getCookie
export const setCookie = cookie.setCookie


export const interpolateString = ( str, callback ) =>
  str.replace( /{{([a-z][a-z0-9\-_.]*)}}/gi , ( matches, replaced ) => callback( replaced ) || '')

// used to flatten nested properties of an object {a: {b: {c: …}}} => 'a.b.c'
export const flattenObject = (obj, roots = [], sep = '.') => Object
  // find props of given object
  .keys(obj)
  // return an object by iterating props
  .reduce((memo, prop) => Object.assign(
    // create a new object
    {},
    // include previously returned object
    memo,
    Object.prototype.toString.call(obj[prop]) === '[object Object]'
      // keep working if value is an object
      ? flattenObject(obj[prop], roots.concat([prop]))
      // include current prop and value and prefix prop with the roots
      : {[roots.concat([prop]).join(sep)]: obj[prop]}
  ), {})

// only used for throttling the 'mousemove' event (used for animating the revoke button when `animateRevokable` is true)
export const throttle = (callback, limit) => {
  let wait = false
  return function() {
    if (!wait) {
      callback(...arguments)
      wait = true
      setTimeout(function() {
        wait = false
      }, limit)
    }
  }
}

// only used for hashing json objects (used for hash mapping palette objects, used when custom colours are passed through JavaScript)
export const hash = str => {
  let hashNum = 0,
    i = 0,
    chr,
    len = str.length
  if (str.length === 0) return hashNum
  for ( i; i < len; ++i ) {
    chr = str.charCodeAt( i )
    hashNum = ( hashNum << 5 ) - hashNum + chr
    hashNum |= 0
  }
  return hashNum
}
import * as style from "./style"
export const normalizeHex = style.normalizeHex
export const getContrast = style.getContrast
export const getLuminance = style.getLuminance
export const getHoverColor =style.getHoverColor

import * as dom from "./dom"
export const traverseDOMPath = dom.traverseDOMPath
export const addCustomStylesheet = dom.addCustomStylesheet

import * as validation from "./validation"
export const isValidStatus = validation.isValidStatus
export const isPlainObject = validation.isPlainObject
export const isMobile = validation.isMobile

import * as asyncFns from "./async"
export const getScript = asyncFns.getScript
export const makeAsyncRequest = asyncFns.makeAsyncRequest

const loopProperties = overwrites => (obj, [key, value]) => {
  if ( value instanceof Object && !(value instanceof Array) ) {
    if ( overwrites[ key ] instanceof Object && !(overwrites[ key ] instanceof Array)) {
      obj[ key ] = Object.entries(value).reduce(loopProperties(overwrites[key]), {})
    } else {
      obj[ key ] = value
    }
  } else {
    if ( overwrites.hasOwnProperty( key ) ) {
      obj[ key ] = overwrites[ key ]
    }else {
      obj[ key ] = value
    }
  }
  return obj
}

export const mergeOptions = ( defaults, overwrites ) =>
  Object.entries(defaults).reduce(loopProperties(overwrites), {})


src/options/popup.js
"use strict"

import { statusDeny, statusAllow, statusDismiss, categories }  from "../constants/index.js"

export default {
  // if false, this prevents the popup from showing (useful for giving to control to another piece of code)
  enabled: true,

  // optional (expecting a HTML element) if passed, the popup is appended to this element. default is `document.body`
  container: null,

  // defaults cookie options - it is RECOMMENDED to set these values to correspond with your server
  cookie: {
    // This is the name of this cookie - you can ignore this
    name: 'cookieconsent_status',
    // This is the url path that the cookie 'name' belongs to. The cookie can only be read at this location
    path: '/',
    // This is the domain that the cookie 'name' belongs to. The cookie can only be read on this domain.
    //  - Guide to cookie domains - https://www.mxsasha.eu/blog/2014/03/04/definitive-guide-to-cookie-domains/
    domain: 'localhost',
    // The cookies expire date, specified in days (specify -1 for no expiry)
    expiryDays: 365,
    // If true the cookie will be created with the secure flag. Secure cookies will only be transmitted via HTTPS.
    secure: false,
    // Sets the "sameSite" attribute of the `cookieconsent_status` cookie
    // - Allowed attributes are "Lax", "Strict" and "None" ("None" is only allowed with the secure flag)
    sameSite: 'Lax',
  },

  // each item defines the inner text for the element that it references
  content: {
    header: 'Cookies used on the website!',
    message: 'This website uses cookies to ensure you get the best experience on our website.',
    policy: 'Cookie Policy',
    link: {
      text: 'Learn more',
      title: 'Learn more about cookies',
      href: 'https://www.cookiesandyou.com',
      target: '_blank'
    },
    buttons: {
      allow: 'Allow cookies',
      close: '&#x274c',
      deny: 'Decline',
      dismiss: 'Got it!',
      save: 'Save',
      tooltip: '^'
    },
    categories: {
      uncategorized: {
        label: 'Uncategorized',
        description: 'This is the category for cookies that don\'t fit any of the other categories'
      },
      essential: {
        label: 'Essential',
        description: 'This is the category for cookies that are essential'
      },
      personalization: {
        label: 'Personalization',
        description: 'This is the category for cookies that are used for personalization'
      },
      analytics: {
        label: 'Analytics',
        description: 'This is the category for tracking cookies'
      },
      marketing: {
        label: 'Marketing',
        description: 'This is the category for marketing cookies'
      },
    },
  },

  // This is the HTML for the elements above. The string {{header}} will be replaced with the equivalent text below.
  // You can remove "{{header}}" and write the content directly inside the HTML if you want.
  //
  //  - ARIA rules suggest to ensure controls are tabbable (so the browser can find the first control),
  //    and to set the focus to the first interactive control (https://w3c.github.io/using-aria/)
  elements: {
    header: '<span class="cc-header">{{header}}</span>&nbsp;',
    message:
      '<span id="cookieconsent:desc" class="cc-message">{{message}}</span>',
    messagelink:
      '<span id="cookieconsent:desc" class="cc-message">{{message}} <a aria-label="learn more about cookies" role=button tabindex="0" class="cc-link" href="{{link.href}}" rel="noopener noreferrer nofollow" target="{{link.target}}" title="{{link.title}}">{{link.text}}</a></span>',
    dismiss:
      `<a aria-label="dismiss cookie message" role=button tabindex="0" class="cc-btn cc-${statusDismiss}">{{buttons.dismiss}}</a>`,
    allow:
      `<a aria-label="allow cookies" role=button tabindex="0" class="cc-btn cc-${statusAllow}">{{buttons.allow}}</a>`,
    deny:
      `<a aria-label="deny cookies" role=button tabindex="0" class="cc-btn cc-${statusDeny}">{{buttons.deny}}</a>`,
    link:
      '<a aria-label="learn more about cookies" role=button tabindex="0" class="cc-link" href="{{link.href}}" rel="noopener noreferrer nofollow" target="{{link.target}}" title="{{link.title}}">{{link.text}}</a>',
    close:
      '<span aria-label="dismiss cookie message" role=button tabindex="0" class="cc-close">{{buttons.close}}</span>',
    categories: '<ul class="cc-categories">' +
      categories.map( ( category, index ) => {
        let categoryLowerCase = category.toLowerCase();
        return `<li class="cc-category">
          <button class="cc-btn" tabindex="0">
            <input type="checkbox" name="${category}" />
            <span class="cc-btn-checkbox"></span>
            {{categories.${categoryLowerCase}.label}}
          </button>
          <button class="cc-btn cc-info" aria-label="${category} Definition Button" tabindex="${index + 1}">
            {{buttons.tooltip}}
          </button>
          <div class="cc-tooltip">
            <p>{{categories.${categoryLowerCase}.description}}</p>
          </div>
        </li>`
      }).join("")
      + '</ul>',
    save: `<button class="cc-btn cc-save">{{buttons.save}}</button>`
    //compliance: compliance is also an element, but it is generated by the application, depending on `type` below
  },

  // The placeholders {{classes}} and {{children}} both get replaced during initialisation:
  //  - {{classes}} is where additional classes get added
  //  - {{children}} is where the HTML children are placed
  window:
    '<div role="dialog" aria-live="polite" aria-label="cookieconsent" aria-describedby="cookieconsent:desc" class="cc-window {{classes}}"><!--googleoff: all-->{{children}}<!--googleon: all--></div>',

  modal: '',

  // This is the html for the revoke button. This only shows up after the user has selected their level of consent
  // It can be enabled of disabled using the `revokable` option
  revokeBtn: '<div class="cc-revoke {{classes}}">{{policy}}</div>',

  // define types of 'compliance' here. '{{value}}' strings in here are linked to `elements`
  compliance: {
    info: '<div class="cc-compliance">{{dismiss}}</div>',
    'opt-in':
      '<div class="cc-compliance cc-highlight">{{dismiss}}{{allow}}{{customize}}</div>',
    'opt-out':
      '<div class="cc-compliance cc-highlight">{{dismiss}}{{deny}}</div>',
    categories: '<div class="form">{{categories}}{{save}}</div>'
  },

  // select your type of popup here
  type: 'info', // refers to `compliance` (in other words, the buttons that are displayed)

  // define layout layouts here
  layouts: {
    // the 'block' layout tend to be for square floating popups
    basic         : '{{messagelink}}{{compliance}}',
    'basic-close' : '{{messagelink}}{{compliance}}{{close}}',
    'basic-header': '{{header}}{{message}}{{link}}{{compliance}}'
    // add a custom layout here, then add some new css with the class '.cc-layout-my-cool-layout'
    //'my-cool-layout': '<div class="my-special-layout">{{message}}{{compliance}}</div>{{close}}',
  },

  // default layout (see above)
  layout: 'basic',

  // this refers to the popup windows position. we currently support:
  //  - banner positions: top, bottom
  //  - floating positions: top-left, top-right, bottom-left, bottom-right
  //
  // adds a class `cc-floating` or `cc-banner` which helps when styling
  position: 'bottom', // default position is 'bottom'

  // Available styles
  //    -block (default, no extra classes)
  //    -edgeless
  //    -classic
  // use your own style name and use `.cc-theme-STYLENAME` class in CSS to edit.
  // Note: style "wire" is used for the configurator, but has no CSS styles of its own, only palette is used.
  theme: 'block',

  // The popup is `fixed` by default, but if you want it to be static (inline with the page content), set this to false
  // Note: by default, we animate the height of the popup from 0 to full size
  static: false,

  // if you want custom colours, pass them in here. this object should look like this.
  // ideally, any custom colours/themes should be created in a separate style sheet, as this is more efficient.
  //   {
  //     popup: {background: '#000000', text: '#fff', link: '#fff'},
  //     button: {background: 'transparent', border: '#f8e71c', text: '#f8e71c'},
  //     highlight: {background: '#f8e71c', border: '#f8e71c', text: '#000000'},
  //   }
  // `highlight` is optional and extends `button`. if it exists, it will apply to the first button
  // only background needs to be defined for every element. if not set, other colors can be calculated from it
  palette: null,

  // Some countries REQUIRE that a user can change their mind. You can configure this yourself.
  // Most of the time this should be false, but the `cookieconsent.legal` can change this to `true` if it detects that it should
  revokable: false,

  // if true, the revokable button will tranlate in and out
  animateRevokable: true,

  // used to disable link on existing layouts
  // replaces element messagelink with message and removes content of link
  showLink: true,

  // set value as scroll range to enable
  dismissOnScroll: false,

  // set value as time in milliseconds to autodismiss after set time
  dismissOnTimeout: false,

  // set value as click anything on the page, excluding the `ignoreClicksFrom` below (if we click on the revoke button etc)
  dismissOnWindowClick: false,

  // set value as click anything on the page, excluding the `ignoreClicksFrom` below (if we click on the revoke button etc)
  dismissOnLinkClick: false,

  // set value as keys are pressed, ( allowKeyCode = 13, denyKeyCode = 27 )
  dismissOnKeyPress: false,

  // If `dismissOnWindowClick` is true, we can click on 'revoke' and we'll still dismiss the banner, so we need exceptions.
  // should be an array of class names (not CSS selectors)
  ignoreClicksFrom: ['cc-revoke', 'cc-btn', 'cc-link'], // already includes the revoke button and the banner itself

  // The application automatically decide whether the popup should open.
  // Set this to false to prevent this from happening and to allow you to control the behaviour yourself
  autoOpen: true,

  // By default the created HTML is automatically appended to the container (which defaults to <body>). You can prevent this behaviour
  // by setting this to false, but if you do, you must attach the `element` yourself, which is a public property of the popup instance:
  //
  //     const instance = cookieconsent.factory(options)
  //     document.body.appendChild(instance.element)
  //
  autoAttach: true,

  // set value if floating layout should be forced for mobile devices
  mobileForceFloat: true,

  // simple whitelist/blacklist for pages. specify page by:
  //   - using a string : '/index.html'           (matches '/index.html' exactly) OR
  //   - using RegExp   : /\/page_[\d]+\.html/    (matched '/page_1.html' and '/page_2.html' etc)
  whitelistPage: [],
  blacklistPage: [],

  // If this is defined, then it is used as the inner html instead of layout. This allows for ultimate customisation.
  // Be sure to use the classes `cc-btn` and `cc-ALLOW`, `cc-DENY` or `cc-DISMISS`. They enable the app to register click
  // handlers. You can use other pre-existing classes too. See `src/styles` folder.
  overrideHTML: null
}